nav2_util/include/nav2_util/service_client.hpp

```diff
@@ -45,22 +70,57 @@ class ServiceClient
     typename RequestType::SharedPtr & request,
     const std::chrono::seconds timeout = std::chrono::seconds::max())
   {
-    auto result_future = client_->async_send_request(request);
+    while (!client_->wait_for_service(std::chrono::seconds(1))) {
+      if (!rclcpp::ok()) {
+        throw std::runtime_error(
+                "ServiceClient: service call interrupted while waiting for service");
+      }
+      RCLCPP_DEBUG(node_->get_logger(), "Waiting for service to appear...");
+    }
+
+    RCLCPP_DEBUG(node_->get_logger(), "send async request");
+    auto future_result = client_->async_send_request(request);
 
-    if (rclcpp::spin_until_future_complete(node_, result_future, timeout) !=
+    if (rclcpp::spin_until_future_complete(node_, future_result, timeout) !=
       rclcpp::executor::FutureReturnCode::SUCCESS)
     {
       throw std::runtime_error("ServiceClient::async_send_request: service call failed");
```

Comment:
Can we add the service name to the error message here too?

Reply:
Fixed.